File: sdk/typescript/src/instructions.ts
/**
 * Instruction Builders
 */
import {
  PublicKey,
  TransactionInstruction,
  SystemProgram,
  SYSVAR_RENT_PUBKEY,
} from '@solana/web3.js';
import { TOKEN_PROGRAM_ID } from '@solana/spl-token';
import BN from 'bn.js';
import {
  ROUTER_PROGRAM_ID,
  SLAB_PROGRAM_ID,
  RouterInstruction,
  SlabInstruction,
  Side,
  TimeInForce,
} from './constants';
import {
  deriveRegistryPda,
  deriveVaultPda,
  derivePortfolioPda,
  deriveInsurancePda,
  deriveVaultAuthorityPda,
} from './pda';
import type {
  OrderParams,
  MultiSlabReserveParams,
  DepositParams,
  WithdrawParams,
  InitializeInsuranceParams,
  ContributeInsuranceParams,
  InitiateWithdrawalParams,
} from './types';

// ============================================================================
// HELPER FUNCTIONS
// ============================================================================

function u8ToBuffer(value: number): Buffer {
  const buf = Buffer.alloc(1);
  buf.writeUInt8(value);
  return buf;
}

function u16ToBuffer(value: number): Buffer {
  const buf = Buffer.alloc(2);
  buf.writeUInt16LE(value);
  return buf;
}

function u64ToBuffer(value: BN): Buffer {
  return value.toArrayLike(Buffer, 'le', 8);
}

function i64ToBuffer(value: BN): Buffer {
  return value.toArrayLike(Buffer, 'le', 8);
}

function u128ToBuffer(value: BN): Buffer {
  return value.toArrayLike(Buffer, 'le', 16);
}

// ============================================================================
// ROUTER INSTRUCTIONS
// ============================================================================

/**
 * Create initialize router instruction
 */
export function createInitializeRouterInstruction(
  admin: PublicKey,
  usdcMint: PublicKey
): TransactionInstruction {
  const [registryPda] = deriveRegistryPda();
  const [vaultPda] = deriveVaultPda();
  const [vaultAuthority] = deriveVaultAuthorityPda();

  const data = Buffer.from([RouterInstruction.Initialize]);

  const keys = [
    { pubkey: registryPda, isSigner: false, isWritable: true },
    { pubkey: vaultPda, isSigner: false, isWritable: true },
    { pubkey: vaultAuthority, isSigner: false, isWritable: false },
    { pubkey: admin, isSigner: true, isWritable: true },
    { pubkey: usdcMint, isSigner: false, isWritable: false },
    { pubkey: SystemProgram.programId, isSigner: false, isWritable: false },
    { pubkey: TOKEN_PROGRAM_ID, isSigner: false, isWritable: false },
    { pubkey: SYSVAR_RENT_PUBKEY, isSigner: false, isWritable: false },
  ];

  return new TransactionInstruction({
    keys,
    programId: ROUTER_PROGRAM_ID,
    data,
  });
}

/**
 * Create initialize portfolio instruction
 */
export function createInitializePortfolioInstruction(
  owner: PublicKey
): TransactionInstruction {
  const [registryPda] = deriveRegistryPda();
  const [portfolioPda] = derivePortfolioPda(owner);

  const data = Buffer.from([RouterInstruction.InitializePortfolio]);

  const keys = [
    { pubkey: registryPda, isSigner: false, isWritable: true },
    { pubkey: portfolioPda, isSigner: false, isWritable: true },
    { pubkey: owner, isSigner: true, isWritable: true },
    { pubkey: SystemProgram.programId, isSigner: false, isWritable: false },
  ];

  return new TransactionInstruction({
    keys,
    programId: ROUTER_PROGRAM_ID,
    data,
  });
}

/**
 * Create deposit instruction
 */
export function createDepositInstruction(
  owner: PublicKey,
  userTokenAccount: PublicKey,
  params: DepositParams
): TransactionInstruction {
  const [registryPda] = deriveRegistryPda();
  const [vaultPda] = deriveVaultPda();
  const [portfolioPda] = derivePortfolioPda(owner);

  const data = Buffer.concat([
    u8ToBuffer(RouterInstruction.Deposit),
    u64ToBuffer(params.amount),
  ]);

  const keys = [
    { pubkey: registryPda, isSigner: false, isWritable: true },
    { pubkey: portfolioPda, isSigner: false, isWritable: true },
    { pubkey: vaultPda, isSigner: false, isWritable: true },
    { pubkey: userTokenAccount, isSigner: false, isWritable: true },
    { pubkey: owner, isSigner: true, isWritable: false },
    { pubkey: TOKEN_PROGRAM_ID, isSigner: false, isWritable: false },
  ];

  return new TransactionInstruction({
    keys,
    programId: ROUTER_PROGRAM_ID,
    data,
  });
}

/**
 * Create withdraw instruction
 */
export function createWithdrawInstruction(
  owner: PublicKey,
  userTokenAccount: PublicKey,
  params: WithdrawParams
): TransactionInstruction {
  const [registryPda] = deriveRegistryPda();
  const [vaultPda] = deriveVaultPda();
  const [vaultAuthority] = deriveVaultAuthorityPda();
  const [portfolioPda] = derivePortfolioPda(owner);

  const data = Buffer.concat([
    u8ToBuffer(RouterInstruction.Withdraw),
    u64ToBuffer(params.amount),
  ]);

  const keys = [
    { pubkey: registryPda, isSigner: false, isWritable: true },
    { pubkey: portfolioPda, isSigner: false, isWritable: true },
    { pubkey: vaultPda, isSigner: false, isWritable: true },
    { pubkey: userTokenAccount, isSigner: false, isWritable: true },
    { pubkey: vaultAuthority, isSigner: false, isWritable: false },
    { pubkey: owner, isSigner: true, isWritable: false },
    { pubkey: TOKEN_PROGRAM_ID, isSigner: false, isWritable: false },
  ];

  return new TransactionInstruction({
    keys,
    programId: ROUTER_PROGRAM_ID,
    data,
  });
}

/**
 * Create multi-slab reserve instruction
 */
export function createMultiSlabReserveInstruction(
  owner: PublicKey,
  slabAccounts: PublicKey[],
  params: MultiSlabReserveParams
): TransactionInstruction {
  const [registryPda] = deriveRegistryPda();
  const [portfolioPda] = derivePortfolioPda(owner);

  // Serialize splits
  const splitBuffers = params.splits.map((split) =>
    Buffer.concat([
      u8ToBuffer(split.slabIndex),
      u8ToBuffer(split.instrumentIndex),
      u64ToBuffer(split.qty),
      u64ToBuffer(split.limitPrice),
    ])
  );

  const data = Buffer.concat([
    u8ToBuffer(RouterInstruction.MultiSlabReserve),
    u8ToBuffer(params.splits.length),
    ...splitBuffers,
    u64ToBuffer(params.totalQty),
    u64ToBuffer(params.requestId),
    u64ToBuffer(params.expiryTs),
  ]);

  const keys = [
    { pubkey: registryPda, isSigner: false, isWritable: true },
    { pubkey: portfolioPda, isSigner: false, isWritable: true },
    { pubkey: owner, isSigner: true, isWritable: false },
    ...slabAccounts.map((pubkey) => ({
      pubkey,
      isSigner: false,
      isWritable: true,
    })),
  ];

  return new TransactionInstruction({
    keys,
    programId: ROUTER_PROGRAM_ID,
    data,
  });
}

/**
 * Create multi-slab commit instruction
 */
export function createMultiSlabCommitInstruction(
  owner: PublicKey,
  slabAccounts: PublicKey[],
  requestId: BN,
  holdIds: BN[]
): TransactionInstruction {
  const [registryPda] = deriveRegistryPda();
  const [portfolioPda] = derivePortfolioPda(owner);

  const data = Buffer.concat([
    u8ToBuffer(RouterInstruction.MultiSlabCommit),
    u64ToBuffer(requestId),
    u8ToBuffer(holdIds.length),
    ...holdIds.map((id) => u64ToBuffer(id)),
  ]);

  const keys = [
    { pubkey: registryPda, isSigner: false, isWritable: true },
    { pubkey: portfolioPda, isSigner: false, isWritable: true },
    { pubkey: owner, isSigner: true, isWritable: false },
    ...slabAccounts.map((pubkey) => ({
      pubkey,
      isSigner: false,
      isWritable: true,
    })),
  ];

  return new TransactionInstruction({
    keys,
    programId: ROUTER_PROGRAM_ID,
    data,
  });
}

/**
 * Create global liquidation instruction
 */
export function createGlobalLiquidationInstruction(
  liquidator: PublicKey,
  targetPortfolio: PublicKey,
  slabAccounts: PublicKey[]
): TransactionInstruction {
  const [registryPda] = deriveRegistryPda();

  const data = Buffer.from([RouterInstruction.GlobalLiquidation]);

  const keys = [
    { pubkey: registryPda, isSigner: false, isWritable: true },
    { pubkey: targetPortfolio, isSigner: false, isWritable: true },
    { pubkey: liquidator, isSigner: true, isWritable: false },
    ...slabAccounts.map((pubkey) => ({
      pubkey,
      isSigner: false,
      isWritable: true,
    })),
  ];

  return new TransactionInstruction({
    keys,
    programId: ROUTER_PROGRAM_ID,
    data,
  });
}

// ============================================================================
// SLAB INSTRUCTIONS
// ============================================================================

/**
 * Create reserve instruction
 */
export function createReserveInstruction(
  slabState: PublicKey,
  router: PublicKey,
  params: OrderParams,
  requestId: BN,
  expiryTs: BN
): TransactionInstruction {
  const data = Buffer.concat([
    u8ToBuffer(SlabInstruction.Reserve),
    u8ToBuffer(params.instrumentIndex),
    u8ToBuffer(params.side),
    u64ToBuffer(params.price),
    u64ToBuffer(params.qty),
    u8ToBuffer(params.timeInForce),
    u64ToBuffer(requestId),
    u64ToBuffer(expiryTs),
  ]);

  const keys = [
    { pubkey: slabState, isSigner: false, isWritable: true },
    { pubkey: router, isSigner: true, isWritable: false },
  ];

  return new TransactionInstruction({
    keys,
    programId: SLAB_PROGRAM_ID,
    data,
  });
}

/**
 * Create commit instruction
 */
export function createCommitInstruction(
  slabState: PublicKey,
  router: PublicKey,
  holdId: BN
): TransactionInstruction {
  const data = Buffer.concat([
    u8ToBuffer(SlabInstruction.Commit),
    u64ToBuffer(holdId),
  ]);

  const keys = [
    { pubkey: slabState, isSigner: false, isWritable: true },
    { pubkey: router, isSigner: true, isWritable: false },
  ];

  return new TransactionInstruction({
    keys,
    programId: SLAB_PROGRAM_ID,
    data,
  });
}

/**
 * Create cancel instruction
 */
export function createCancelInstruction(
  slabState: PublicKey,
  router: PublicKey,
  holdId: BN
): TransactionInstruction {
  const data = Buffer.concat([
    u8ToBuffer(SlabInstruction.Cancel),
    u64ToBuffer(holdId),
  ]);

  const keys = [
    { pubkey: slabState, isSigner: false, isWritable: true },
    { pubkey: router, isSigner: true, isWritable: false },
  ];

  return new TransactionInstruction({
    keys,
    programId: SLAB_PROGRAM_ID,
    data,
  });
}

// ============================================================================
// INSURANCE INSTRUCTIONS
// ============================================================================

/**
 * Create initialize insurance instruction
 */
export function createInitializeInsuranceInstruction(
  slabState: PublicKey,
  lpOwner: PublicKey,
  params: InitializeInsuranceParams
): TransactionInstruction {
  const [insurancePda] = deriveInsurancePda(slabState);

  const data = Buffer.concat([
    u8ToBuffer(SlabInstruction.InitializeInsurance),
    u64ToBuffer(new BN(params.contributionRateBps)),
    u64ToBuffer(new BN(params.adlThresholdBps)),
    u64ToBuffer(new BN(params.withdrawalTimelockSecs)),
  ]);

  const keys = [
    { pubkey: slabState, isSigner: false, isWritable: true },
    { pubkey: insurancePda, isSigner: false, isWritable: true },
    { pubkey: lpOwner, isSigner: true, isWritable: true },
    { pubkey: SystemProgram.programId, isSigner: false, isWritable: false },
  ];

  return new TransactionInstruction({
    keys,
    programId: SLAB_PROGRAM_ID,
    data,
  });
}

/**
 * Create contribute insurance instruction
 */
export function createContributeInsuranceInstruction(
  slabState: PublicKey,
  lpOwner: PublicKey,
  lpTokenAccount: PublicKey,
  insuranceVault: PublicKey,
  params: ContributeInsuranceParams
): TransactionInstruction {
  const [insurancePda] = deriveInsurancePda(slabState);

  const data = Buffer.concat([
    u8ToBuffer(SlabInstruction.ContributeInsurance),
    u64ToBuffer(params.amount),
  ]);

  const keys = [
    { pubkey: insurancePda, isSigner: false, isWritable: true },
    { pubkey: lpTokenAccount, isSigner: false, isWritable: true },
    { pubkey: insuranceVault, isSigner: false, isWritable: true },
    { pubkey: lpOwner, isSigner: true, isWritable: false },
    { pubkey: TOKEN_PROGRAM_ID, isSigner: false, isWritable: false },
  ];

  return new TransactionInstruction({
    keys,
    programId: SLAB_PROGRAM_ID,
    data,
  });
}

/**
 * Create initiate insurance withdrawal instruction
 */
export function createInitiateInsuranceWithdrawalInstruction(
  slabState: PublicKey,
  lpOwner: PublicKey,
  params: InitiateWithdrawalParams
): TransactionInstruction {
  const [insurancePda] = deriveInsurancePda(slabState);

  const data = Buffer.concat([
    u8ToBuffer(SlabInstruction.InitiateInsuranceWithdrawal),
    u64ToBuffer(params.amount),
  ]);

  const keys = [
    { pubkey: insurancePda, isSigner: false, isWritable: true },
    { pubkey: lpOwner, isSigner: true, isWritable: false },
  ];

  return new TransactionInstruction({
    keys,
    programId: SLAB_PROGRAM_ID,
    data,
  });
}

/**
 * Create complete insurance withdrawal instruction
 */
export function createCompleteInsuranceWithdrawalInstruction(
  slabState: PublicKey,
  lpOwner: PublicKey,
  lpTokenAccount: PublicKey,
  insuranceVault: PublicKey,
  vaultAuthority: PublicKey
): TransactionInstruction {
  const [insurancePda] = deriveInsurancePda(slabState);

  const data = Buffer.from([SlabInstruction.CompleteInsuranceWithdrawal]);

  const keys = [
    { pubkey: insurancePda, isSigner: false, isWritable: true },
    { pubkey: insuranceVault, isSigner: false, isWritable: true },
    { pubkey: lpTokenAccount, isSigner: false, isWritable: true },
    { pubkey: lpOwner, isSigner: true, isWritable: false },
    { pubkey: vaultAuthority, isSigner: false, isWritable: false },
    { pubkey: TOKEN_PROGRAM_ID, isSigner: false, isWritable: false },
  ];

  return new TransactionInstruction({
    keys,
    programId: SLAB_PROGRAM_ID,
    data,
  });
}

/**
 * Create cancel insurance withdrawal instruction
 */
export function createCancelInsuranceWithdrawalInstruction(
  slabState: PublicKey,
  lpOwner: PublicKey
): TransactionInstruction {
  const [insurancePda] = deriveInsurancePda(slabState);

  const data = Buffer.from([SlabInstruction.CancelInsuranceWithdrawal]);

  const keys = [
    { pubkey: insurancePda, isSigner: false, isWritable: true },
    { pubkey: lpOwner, isSigner: true, isWritable: false },
  ];

  return new TransactionInstruction({
    keys,
    programId: SLAB_PROGRAM_ID,
    data,
  });
}

File: sdk/rust/src/instructions.rs
//! Instruction builders

use solana_sdk::{
    instruction::{AccountMeta, Instruction},
    pubkey::Pubkey,
    system_program,
    sysvar,
};
use crate::constants::*;
use crate::pda::*;
use crate::types::*;

// ============================================================================
// ROUTER INSTRUCTIONS
// ============================================================================

/// Create initialize router instruction
pub fn create_initialize_router_instruction(admin: &Pubkey, usdc_mint: &Pubkey) -> Instruction {
    let (registry_pda, _) = derive_registry_pda();
    let (vault_pda, _) = derive_vault_pda();
    let (vault_authority, _) = derive_vault_authority_pda();

    let data = vec![RouterInstruction::Initialize as u8];

    let accounts = vec![
        AccountMeta::new(registry_pda, false),
        AccountMeta::new(vault_pda, false),
        AccountMeta::new_readonly(vault_authority, false),
        AccountMeta::new(*admin, true),
        AccountMeta::new_readonly(*usdc_mint, false),
        AccountMeta::new_readonly(system_program::ID, false),
        AccountMeta::new_readonly(spl_token::ID, false),
        AccountMeta::new_readonly(sysvar::rent::ID, false),
    ];

    Instruction::new_with_bytes(ROUTER_PROGRAM_ID, &data, accounts)
}

/// Create initialize portfolio instruction
pub fn create_initialize_portfolio_instruction(owner: &Pubkey) -> Instruction {
    let (registry_pda, _) = derive_registry_pda();
    let (portfolio_pda, _) = derive_portfolio_pda(owner);

    let data = vec![RouterInstruction::InitializePortfolio as u8];

    let accounts = vec![
        AccountMeta::new(registry_pda, false),
        AccountMeta::new(portfolio_pda, false),
        AccountMeta::new(*owner, true),
        AccountMeta::new_readonly(system_program::ID, false),
    ];

    Instruction::new_with_bytes(ROUTER_PROGRAM_ID, &data, accounts)
}

/// Create deposit instruction
pub fn create_deposit_instruction(
    owner: &Pubkey,
    user_token_account: &Pubkey,
    params: &DepositParams,
) -> Instruction {
    let (registry_pda, _) = derive_registry_pda();
    let (vault_pda, _) = derive_vault_pda();
    let (portfolio_pda, _) = derive_portfolio_pda(owner);

    let mut data = vec![RouterInstruction::Deposit as u8];
    data.extend_from_slice(&params.amount.to_le_bytes());

    let accounts = vec![
        AccountMeta::new(registry_pda, false),
        AccountMeta::new(portfolio_pda, false),
        AccountMeta::new(vault_pda, false),
        AccountMeta::new(*user_token_account, false),
        AccountMeta::new_readonly(*owner, true),
        AccountMeta::new_readonly(spl_token::ID, false),
    ];

    Instruction::new_with_bytes(ROUTER_PROGRAM_ID, &data, accounts)
}

/// Create withdraw instruction
pub fn create_withdraw_instruction(
    owner: &Pubkey,
    user_token_account: &Pubkey,
    params: &WithdrawParams,
) -> Instruction {
    let (registry_pda, _) = derive_registry_pda();
    let (vault_pda, _) = derive_vault_pda();
    let (vault_authority, _) = derive_vault_authority_pda();
    let (portfolio_pda, _) = derive_portfolio_pda(owner);

    let mut data = vec![RouterInstruction::Withdraw as u8];
    data.extend_from_slice(&params.amount.to_le_bytes());

    let accounts = vec![
        AccountMeta::new(registry_pda, false),
        AccountMeta::new(portfolio_pda, false),
        AccountMeta::new(vault_pda, false),
        AccountMeta::new(*user_token_account, false),
        AccountMeta::new_readonly(vault_authority, false),
        AccountMeta::new_readonly(*owner, true),
        AccountMeta::new_readonly(spl_token::ID, false),
    ];

    Instruction::new_with_bytes(ROUTER_PROGRAM_ID, &data, accounts)
}

/// Create multi-slab reserve instruction
pub fn create_multi_slab_reserve_instruction(
    owner: &Pubkey,
    slab_accounts: &[Pubkey],
    params: &MultiSlabReserveParams,
) -> Instruction {
    let (registry_pda, _) = derive_registry_pda();
    let (portfolio_pda, _) = derive_portfolio_pda(owner);

    let mut data = vec![RouterInstruction::MultiSlabReserve as u8];
    data.push(params.splits.len() as u8);

    for split in &params.splits {
        data.push(split.slab_index);
        data.push(split.instrument_index);
        data.extend_from_slice(&split.qty.to_le_bytes());
        data.extend_from_slice(&split.limit_price.to_le_bytes());
    }

    data.extend_from_slice(&params.total_qty.to_le_bytes());
    data.extend_from_slice(&params.request_id.to_le_bytes());
    data.extend_from_slice(&params.expiry_ts.to_le_bytes());

    let mut accounts = vec![
        AccountMeta::new(registry_pda, false),
        AccountMeta::new(portfolio_pda, false),
        AccountMeta::new_readonly(*owner, true),
    ];

    for slab in slab_accounts {
        accounts.push(AccountMeta::new(*slab, false));
    }

    Instruction::new_with_bytes(ROUTER_PROGRAM_ID, &data, accounts)
}

/// Create multi-slab commit instruction
pub fn create_multi_slab_commit_instruction(
    owner: &Pubkey,
    slab_accounts: &[Pubkey],
    request_id: u64,
    hold_ids: &[u64],
) -> Instruction {
    let (registry_pda, _) = derive_registry_pda();
    let (portfolio_pda, _) = derive_portfolio_pda(owner);

    let mut data = vec![RouterInstruction::MultiSlabCommit as u8];
    data.extend_from_slice(&request_id.to_le_bytes());
    data.push(hold_ids.len() as u8);

    for hold_id in hold_ids {
        data.extend_from_slice(&hold_id.to_le_bytes());
    }

    let mut accounts = vec![
        AccountMeta::new(registry_pda, false),
        AccountMeta::new(portfolio_pda, false),
        AccountMeta::new_readonly(*owner, true),
    ];

    for slab in slab_accounts {
        accounts.push(AccountMeta::new(*slab, false));
    }

    Instruction::new_with_bytes(ROUTER_PROGRAM_ID, &data, accounts)
}

/// Create global liquidation instruction
pub fn create_global_liquidation_instruction(
    liquidator: &Pubkey,
    target_portfolio: &Pubkey,
    slab_accounts: &[Pubkey],
) -> Instruction {
    let (registry_pda, _) = derive_registry_pda();

    let data = vec![RouterInstruction::GlobalLiquidation as u8];

    let mut accounts = vec![
        AccountMeta::new(registry_pda, false),
        AccountMeta::new(*target_portfolio, false),
        AccountMeta::new_readonly(*liquidator, true),
    ];

    for slab in slab_accounts {
        accounts.push(AccountMeta::new(*slab, false));
    }

    Instruction::new_with_bytes(ROUTER_PROGRAM_ID, &data, accounts)
}

// ============================================================================
// SLAB INSTRUCTIONS
// ============================================================================

/// Create reserve instruction
pub fn create_reserve_instruction(
    slab_state: &Pubkey,
    router: &Pubkey,
    params: &OrderParams,
    request_id: u64,
    expiry_ts: u64,
) -> Instruction {
    let mut data = vec![SlabInstruction::Reserve as u8];
    data.push(params.instrument_index);
    data.push(params.side as u8);
    data.extend_from_slice(&params.price.to_le_bytes());
    data.extend_from_slice(&params.qty.to_le_bytes());
    data.push(params.time_in_force as u8);
    data.extend_from_slice(&request_id.to_le_bytes());
    data.extend_from_slice(&expiry_ts.to_le_bytes());

    let accounts = vec![
        AccountMeta::new(*slab_state, false),
        AccountMeta::new_readonly(*router, true),
    ];

    Instruction::new_with_bytes(SLAB_PROGRAM_ID, &data, accounts)
}

/// Create commit instruction
pub fn create_commit_instruction(
    slab_state: &Pubkey,
    router: &Pubkey,
    hold_id: u64,
) -> Instruction {
    let mut data = vec![SlabInstruction::Commit as u8];
    data.extend_from_slice(&hold_id.to_le_bytes());

    let accounts = vec![
        AccountMeta::new(*slab_state, false),
        AccountMeta::new_readonly(*router, true),
    ];

    Instruction::new_with_bytes(SLAB_PROGRAM_ID, &data, accounts)
}

/// Create cancel instruction
pub fn create_cancel_instruction(
    slab_state: &Pubkey,
    router: &Pubkey,
    hold_id: u64,
) -> Instruction {
    let mut data = vec![SlabInstruction::Cancel as u8];
    data.extend_from_slice(&hold_id.to_le_bytes());

    let accounts = vec![
        AccountMeta::new(*slab_state, false),
        AccountMeta::new_readonly(*router, true),
    ];

    Instruction::new_with_bytes(SLAB_PROGRAM_ID, &data, accounts)
}

// ============================================================================
// INSURANCE INSTRUCTIONS
// ============================================================================

/// Create initialize insurance instruction
pub fn create_initialize_insurance_instruction(
    slab_state: &Pubkey,
    lp_owner: &Pubkey,
    params: &InitializeInsuranceParams,
) -> Instruction {
    let (insurance_pda, _) = derive_insurance_pda(slab_state);

    let mut data = vec![SlabInstruction::InitializeInsurance as u8];
    data.extend_from_slice(&params.contribution_rate_bps.to_le_bytes());
    data.extend_from_slice(&params.adl_threshold_bps.to_le_bytes());
    data.extend_from_slice(&params.withdrawal_timelock_secs.to_le_bytes());

    let accounts = vec![
        AccountMeta::new(*slab_state, false),
        AccountMeta::new(insurance_pda, false),
        AccountMeta::new(*lp_owner, true),
        AccountMeta::new_readonly(system_program::ID, false),
    ];

    Instruction::new_with_bytes(SLAB_PROGRAM_ID, &data, accounts)
}

/// Create contribute insurance instruction
pub fn create_contribute_insurance_instruction(
    slab_state: &Pubkey,
    lp_owner: &Pubkey,
    lp_token_account: &Pubkey,
    insurance_vault: &Pubkey,
    params: &ContributeInsuranceParams,
) -> Instruction {
    let (insurance_pda, _) = derive_insurance_pda(slab_state);

    let mut data = vec![SlabInstruction::ContributeInsurance as u8];
    data.extend_from_slice(&params.amount.to_le_bytes());

    let accounts = vec![
        AccountMeta::new(insurance_pda, false),
        AccountMeta::new(*lp_token_account, false),
        AccountMeta::new(*insurance_vault, false),
        AccountMeta::new_readonly(*lp_owner, true),
        AccountMeta::new_readonly(spl_token::ID, false),
    ];

    Instruction::new_with_bytes(SLAB_PROGRAM_ID, &data, accounts)
}

/// Create initiate insurance withdrawal instruction
pub fn create_initiate_insurance_withdrawal_instruction(
    slab_state: &Pubkey,
    lp_owner: &Pubkey,
    params: &InitiateWithdrawalParams,
) -> Instruction {
    let (insurance_pda, _) = derive_insurance_pda(slab_state);

    let mut data = vec![SlabInstruction::InitiateInsuranceWithdrawal as u8];
    data.extend_from_slice(&params.amount.to_le_bytes());

    let accounts = vec![
        AccountMeta::new(insurance_pda, false),
        AccountMeta::new_readonly(*lp_owner, true),
    ];

    Instruction::new_with_bytes(SLAB_PROGRAM_ID, &data, accounts)
}

/// Create complete insurance withdrawal instruction
pub fn create_complete_insurance_withdrawal_instruction(
    slab_state: &Pubkey,
    lp_owner: &Pubkey,
    lp_token_account: &Pubkey,
    insurance_vault: &Pubkey,
    vault_authority: &Pubkey,
) -> Instruction {
    let (insurance_pda, _) = derive_insurance_pda(slab_state);

    let data = vec![SlabInstruction::CompleteInsuranceWithdrawal as u8];

    let accounts = vec![
        AccountMeta::new(insurance_pda, false),
        AccountMeta::new(*insurance_vault, false),
        AccountMeta::new(*lp_token_account, false),
        AccountMeta::new_readonly(*lp_owner, true),
        AccountMeta::new_readonly(*vault_authority, false),
        AccountMeta::new_readonly(spl_token::ID, false),
    ];

    Instruction::new_with_bytes(SLAB_PROGRAM_ID, &data, accounts)
}

/// Create cancel insurance withdrawal instruction
pub fn create_cancel_insurance_withdrawal_instruction(
    slab_state: &Pubkey,
    lp_owner: &Pubkey,
) -> Instruction {
    let (insurance_pda, _) = derive_insurance_pda(slab_state);

    let data = vec![SlabInstruction::CancelInsuranceWithdrawal as u8];

    let accounts = vec![
        AccountMeta::new(insurance_pda, false),
        AccountMeta::new_readonly(*lp_owner, true),
    ];

    Instruction::new_with_bytes(SLAB_PROGRAM_ID, &data, accounts)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_deposit_instruction() {
        let owner = Pubkey::new_unique();
        let token_account = Pubkey::new_unique();
        let params = DepositParams { amount: 1_000_000 };

        let ix = create_deposit_instruction(&owner, &token_account, &params);

        assert_eq!(ix.program_id, ROUTER_PROGRAM_ID);
        assert!(!ix.data.is_empty());
        assert_eq!(ix.data[0], RouterInstruction::Deposit as u8);
    }

    #[test]
    fn test_reserve_instruction() {
        let slab = Pubkey::new_unique();
        let router = Pubkey::new_unique();
        let params = OrderParams {
            instrument_index: 0,
            side: Side::Buy,
            price: 50_000_000_000,
            qty: 1_000_000,
            time_in_force: TimeInForce::GTC,
            ..Default::default()
        };

        let ix = create_reserve_instruction(&slab, &router, &params, 1, 0);

        assert_eq!(ix.program_id, SLAB_PROGRAM_ID);
        assert_eq!(ix.data[0], SlabInstruction::Reserve as u8);
    }
}
